### WRANGLE AND ANALYZE DATA PROJECT

### Table of Contents

* [Introduction](#Introdcution)
* [Data Gathering](#Data-Gathering)
* [Data Assessment](#Data-Assessement)
* [Data Cleaning](#Data-Cleaning)

#### Introduction

The aim of this project is to wrangle WeRateDogs Twitter data in order to create interesting and trustworthy analyses and visualisations.  
The data wrangling process consist of three phases 

    - Gathering data
    - Assessing data
    - Cleaning data
    
- The cleaned data will be analysed and used to create visuals to give possible interpretation for the data. A report for both processes will also be submitted at the end. 

#### Gathering Data
Three pieces of data is required for this project:

1. The WeRateDogs Twitter archive data, provided by Udacity and downloaded manually from Udacity resource center
2. The tweet image predictions, hosted on Udacity's server and downloaded programmatically using `Get Requests` 
3. Retweet and favorite counts for the tweet_id's in the archived data from `1` above. This data is accessed by qerying the Twitter API for each tweet's JSON data using Python's Tweepy library and store each tweet's entire set of JSON data in a file called tweet_json.txt file.

#### Assessing Data 
- After gathering each of the above pieces of data, assess them visually and programmatically for quality and tidiness issues.
- Target is to  detect and document at least eight (8) quality issues and two (2) tidiness issues.

#### Cleaning Data
- Issues documented while assessing wil be cleaned to give an output of high quality and tidy master pandas DataFrame.

#### Storing, Analyzing, and Visualizing Data 
- Store the clean DataFrame(s) in a CSV file with the main one named `twitter_archive_master.csv`. Wrangled data will be analysed and visualised using Jupyter notebook. This should include at least three (3) insights and one (1) visualisation. 
- A written report will prepared to describe the wrangling efforts and second report to communicate the insights and displays will also be prepared. 

In [1]:
#Import Necessary modules
import pandas as pd
import numpy as np
import os
#to make requests
import requests
#to display tables
from IPython.display import display
#to access twitter APi
import tweepy as tw

#to write json to pandas dataframe
from pandas import DataFrame

#for json file
import json


In [2]:
%%HTML
<style type="text/css">
table.dataframe td, table.dataframe th {
    border: 1px  black solid !important;
  color: black !important;
}
</style> 

### DATA GATHERING

Three pieces of data will be gathered for this project.

**Data One** 

The WeRateDogs Twitter archive. Provided and made available by Udacity. Downloaded from the resource centre and loaded into notebook as **.csv**

In [3]:
#import data to have an overview
df1 = pd.read_csv('twitter_archive_enhanced.csv')

In [4]:
df1.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None

In [5]:
df1.shape

(2356, 17)

**Data Two**

The tweet image predictions, this information using a neural network predicts what breed of dog (or other object, animal, etc.) is present in each tweet. This was already provided by Udacity.

- This file (image_predictions.tsv) is hosted on Udacity's servers and downloaded using **Requests** library

In [6]:
#Getting the url
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)

In [7]:
#to view raw data
#response.content

In [53]:
#savingg file to computer
with open('C:/Users/Frances-Anthony/Documents/Udacity/data_wrangle_analyze_project/image_predictions.tsv', mode='wb') as file:
    file.write(response.content)

In [6]:
#Load image data
df2 = pd.read_csv('image_predictions.tsv', delimiter="\t")
df2.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

**Data Three** 

Extract each tweet's retweet count and favorite ("like") count for the `tweet_id` in the archived data downloaded. 

**HOW:** 
1. using the tweet IDs in the `WeRateDogs` Twitter archive, 
2. query the Twitter API for each tweet's JSON data using Python's `Tweepy library` and 
3. store each tweet's entire set of JSON data in a file called `tweet_json.txt file.`
4. Read the `.txt` file line by line into a pandas frame with these columns `(tweet_id, retweet_count, favorite_count).`

In [ ]:
'''
Code source:
#https://www.earthdatascience.org/courses/use-data-open-source-python/intro-to-apis/twitter-data-in-python/
#https://towardsdatascience.com/data-analysis-project-weratedogs-16aad3f6c8c2
'''
#import necessary modules in this insatnce Tweepy has already been imported above
#from the twitter archved data provided by Udacity call tweet_id to query twitter API for retweets and "likes" counts
#to access twitter API, input consumer and secret key gotten from twitter
#for privacy, i will leave them blank

#define keys
consumer_key= '.......'
consumer_secret = '..........'
access_token = '..........'
access_token_secret = '........'

auth = tw.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)
api = tw.API(auth, wait_on_rate_limit=True)

#get tweets from twitter API 
retweet_favorite_count = []

#save missing tweets in this list
not_found = []

with open('tweet_json.txt', mode = "w") as file:
    for i in list(df1.tweet_id):
        try:
            tweet = api.get_status(str(i))
            file.write(json.dumps(tweet._json))
            retweet_favorite_count.append({"tweet_id":str(i),
                                          "retweet_count": tweet._json['retweet_count'],
                                          "favourite_count": tweet._json['favorite_count']})
        except:
                not_found.append(i)

In [90]:
#Read the .txt file line by line into a pandas frame with these columns 
#(tweet_id, retweet_count, favorite count).
#df3 = pd.DataFrame(retweet_favorite_count, columns=['tweet_id', 'retweet_count', 'favourite_count'])

In [66]:
#write to a csv file
#df3.to_csv('retweet_and_favorites_counts.csv')

In [7]:
#read retweet_favorite_counts as csv and assign to dataframe
df3 = pd.read_csv('retweet_and_favorite_counts.csv')
df3.head()

Unnamed: 0            tweet_id  retweet_count  favourite_count
0           0  892420643555336193           7340            34978
1           1  892177421306343426           5475            30279
2           2  891815181378084864           3621            22784
3           3  891689557279858688           7529            38245
4           4  891327558926688256           8108            36522

In [8]:
df3.shape

(2331, 4)

## ASSESSING


The next phase of the project is assessing the gathered data. I will be assessing the data for quality and tidiness both manually and programmatically using pandas methods.


#### 1. ASSESSING TWITTER ARCHIVE DATA

In [9]:
#renamed the df 
archived_data = df1
archived_data.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

#### COLUMN DESCRIPTION
- tweet_id: unique tweet identifier
- in_reply_to_status_id:
- in_reply_to_user_id:
- timestamp: time of the tweet
- source: where the tweet originated
- text: dog caption
- retweeted_status_id: status identifier for retweets
- retweeted_status_user_id: user identifier for retweets
- retweeted_status_timestamp: time of retweet
- expanded_urls: the url where the tweet is housed
- rating_numerator: rated on a scale of 1-10, but most have ratings above the max of 10
- rating_denominator: usually 10 (original maximum)
- name: given name of the dog
- doggo: dog stage (adult)
- floofer: dog stage (fluffy)
- pupper: dog stage (young)
- puppo: dog stage (transitioning from young to adult)

### QUALITY ISSUES
- missing data: in_reply_to_status_id and in_reply_to_user_id has only 78 rows available out of the 2356.
- not original tweet, columns retweeted_status_timestamp, retweeted status_id and user_id means that rows are retweet and not original tweet. (drop rows)	
- dog name is none for 745 rows - check if the same rows with retweet
- dog names recorded as `a` or `an` should be `None`
- rows with `a` also have dog style
- does not contain retweet and favorite counts 
- inconsistent rating_denominator -value should be 10, values    greater than 10 should be removed
- very high rating_numerator as much 1776, not an issue but keep in mind
- time stamp has object data type, change to datetime
- some rows have several identical values in the expanded_url column concatenated by a comma.

### TIDINESS ISSUES
- create one column for dog stage, collapse multiple colunmns and rows. convert the rows (name, doggo, floofer,pupper,puppo into two columns, one with dog name and one with dog level specifying either from the list)
- expanded url has multiple url on one row
- the three datasets can be one,all have tweet_id

In [10]:
archived_data.shape

(2356, 17)

In [11]:
archived_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [12]:
archived_data.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [13]:
archived_data.puppo.unique()

array(['None', 'puppo'], dtype=object)

In [14]:
archived_data.puppo.value_counts()

None     2326
puppo      30
Name: puppo, dtype: int64

In [15]:
archived_data.retweeted_status_id.nunique() 

181

In [16]:
#archived_data.rating_denominator.isnull()
archived_data.rating_numerator.isnull().sum().any()

False

In [17]:
archived_data.name.nunique(),archived_data.name.value_counts();

In [18]:
#check for duplicates
archived_data.tweet_id.duplicated().sum()

0

In [19]:
archived_data.tweet_id.nunique() #no duplicates

2356

In [20]:
#archived_data.name.isnull().any()
archived_data.query('name == "None"');

In [21]:
#dog names recorded as a or an
#archived_data.query('name == "an"')
archived_data.query('name == "a"');

In [22]:
archived_data.retweeted_status_id.notnull().sum()

181

In [23]:
archived_data.rating_denominator.nunique(),archived_data.rating_denominator.duplicated().sum()

(18, 2338)

In [24]:
#archived_data.rating_denominator.value_counts()
archived_data.rating_denominator.isnull().sum().any()

False

In [25]:
archived_data.rating_denominator.unique()

array([ 10,   0,  15,  70,   7,  11, 150, 170,  20,  50,  90,  80,  40,
       130, 110,  16, 120,   2], dtype=int64)

In [26]:
archived_data.rating_denominator.value_counts();

In [27]:
#archived_data.rating_numerator.notnull().sum()
archived_data.rating_numerator.isnull().sum()

0

In [37]:
archived_data.rating_numerator.unique();

In [36]:
archived_data.rating_numerator.value_counts();

In [30]:
archived_data.query('rating_numerator == 1776')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
979  749981277374128128                    NaN                  NaN   

                     timestamp  \
979  2016-07-04 15:00:45 +0000   

                                                source  \
979  <a href="https://about.twitter.com/products/tw...   

                                                  text  retweeted_status_id  \
979  This is Atticus. He's quite simply America af....                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
979                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
979  https://twitter.com/dog_rates/status/749981277...              1776   

     rating_denominator     name doggo floofer pupper puppo  
979                  10  Atticus  None    None   None  None

In [31]:
archived_data.rating_numerator.isnull().sum()

0

In [33]:
archived_data.expanded_urls.isnull().sum()

59

#### 2. ASSESSING IMAGE PREDICTIONS DATA

In [38]:
#Reassign dataframe
image_pred = df2
image_pred.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [39]:
image_pred.shape

(2075, 12)

#### Column Description

- `tweet_id` is the last part of the tweet URL after "status/" 
- `p1` is the algorithm's #1 prediction for the image in the tweet
- `p1_conf` is how confident the algorithm is in its #1 prediction in `%`
- `p1_dog` is whether or not the #1 prediction is a breed of dog
- `p2` is the algorithm's second most likely prediction
- `p2_conf` is how confident the algorithm is in its #2 prediction 
- `p2_dog` is whether or not the #2 prediction is a breed of dog
- `p3` is the algorithm's 3rd most likely prediction
- `p3_conf` is how confident the algorithm is in its #3 prediction 
- `p3_dog` is whether or not the #3 prediction is a breed of dog

In [40]:
image_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


#### Quality issues
- inconsistent naming for dog breeds p1, remove underscore
- convert prediction number to type `int` and remove letter `p` 
- inconsistent labelling for dog breeds, convert all breed name to lower case
- some breed predictions are false
#### Tidiness issues
- Predictions are spread in three columns.
- Confidence intervals are spread in three columns.
- Dog tests are spread in three columns.
- Melt all three into two columns (breed and confidence)

In [41]:
sum(image_pred.tweet_id.duplicated())

0

In [42]:
image_pred.p1_dog.unique(), image_pred.p2_dog.unique(), image_pred.p3_dog.unique()

(array([ True, False]), array([ True, False]), array([ True, False]))

In [43]:
image_pred.query('p1 == "NaN"')

Empty DataFrame
Columns: [tweet_id, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog]
Index: []

In [400]:
image_pred.p1.unique();

#### 3. ASSESSING RETWEET AND FAVORITE COUNT DATA

In [44]:
rtwt_fav_count = df3
rtwt_fav_count.head()

Unnamed: 0            tweet_id  retweet_count  favourite_count
0           0  892420643555336193           7340            34978
1           1  892177421306343426           5475            30279
2           2  891815181378084864           3621            22784
3           3  891689557279858688           7529            38245
4           4  891327558926688256           8108            36522

#### Column Description
- `tweet_id` the tweet_id
- `retweet_count` number of retweet for each Id
- `favourite_count` number of `likes` for the Id

In [45]:
rtwt_fav_count.shape

(2331, 4)

In [46]:
rtwt_fav_count.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   Unnamed: 0       2331 non-null   int64
 1   tweet_id         2331 non-null   int64
 2   retweet_count    2331 non-null   int64
 3   favourite_count  2331 non-null   int64
dtypes: int64(4)
memory usage: 73.0 KB


#### Quality issues
- first column(unamed 0) not needed - drop

#### Tidiness issues
- None

In [47]:
rtwt_fav_count.isnull().sum().any()

False

### CLEANING

**CLEANING FOR ARCHIVED DATA**

In [48]:
df1 = pd.read_csv('twitter_archive_enhanced.csv')

In [49]:
archived_data = df1

In [50]:
#Make copies of dataframe
archived_data_clean = archived_data
image_pred_clean = image_pred
rtwt_fav_count_clean = rtwt_fav_count

In [51]:
archived_data_clean.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None

<font color='red'> **TIDINESS ISSUE - 1** </font>

**Define**

- Name and dog stages row do not obey [Tidy Data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html) rule.
- Use pandas method `assign`: To melt the dog stages columns that occured in multiple columns **doggo, floofer, pupper, puppo** to obey the tidy data rule. 

**CODE**

In [66]:
#create a copy for df
archived_data_clean_melt = archived_data_clean

In [67]:
#Source:https://towardsdatascience.com/data-analysis-project-weratedogs-16aad3f6c8c2
#count for values in individual columns
archived_data_clean_melt['dog_stage'] = None
archived_data_clean_melt['dog_stage'] = archived_data_clean_melt.doggo + archived_data_clean_melt.floofer + archived_data_clean_melt.pupper + archived_data_clean_melt.puppo
archived_data_clean_melt['dog_stage'].value_counts()

NoneNoneNoneNone        1976
NoneNonepupperNone       245
doggoNoneNoneNone         83
NoneNoneNonepuppo         29
doggoNonepupperNone       12
NoneflooferNoneNone        9
doggoflooferNoneNone       1
doggoNoneNonepuppo         1
Name: dog_stage, dtype: int64

In [68]:
#use assign method to reassign the vlaues to the new column called dog_stage
archived_data_clean_melt = archived_data_clean_melt.assign(dog_stage = archived_data_clean_melt.doggo.astype(str) + archived_data_clean_melt.floofer.astype(str) + 
archived_data_clean_melt.pupper.astype(str) + archived_data_clean_melt.puppo.astype(str))                                             

In [69]:
archived_data_clean_melt.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo         dog_stage  
0                  10   Phineas  None    None   None  None  NoneNoneNoneNone  
1                  10     Tilly  None    None   None  None  NoneNoneNoneNone  
2                  10    Archie  None    None   None  None  NoneNoneNoneNone  
3                  10     Darla  None    None   None  None  NoneNoneNoneNone  
4                  10  Franklin  None    None   None  None  NoneNoneNoneNone

In [70]:
#replace all None with empty
archived_data_clean_melt['dog_stage'] = archived_data_clean_melt['dog_stage'].map(lambda x: x.replace("None",""))

In [71]:
archived_data_clean_melt.sample(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1126  729854734790754305                    NaN                  NaN   
1487  693095443459342336                    NaN                  NaN   
598   798686750113755136                    NaN                  NaN   

                      timestamp  \
1126  2016-05-10 02:05:03 +0000   
1487  2016-01-29 15:36:45 +0000   
598   2016-11-16 00:38:54 +0000   

                                                 source  \
1126  <a href="http://twitter.com/download/iphone" r...   
1487  <a href="http://twitter.com/download/iphone" r...   
598   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1126  Say hello to Ollie. He conducts this train. He...                  NaN   
1487  This is Milo. He doesn't understand your fancy...                  NaN   
598   RT @dog_rates: Say hello to Jack (pronounced "...         6.833919e+17   

      retweeted_status_user_id retweeted_status_timestamp  \
1126                       NaN                        NaN   
1487                       NaN                        NaN   
598               4.196984e+09  2016-01-02 20:58:09 +0000   

                                          expanded_urls  rating_numerator  \
1126  https://twitter.com/dog_rates/status/729854734...                11   
1487  https://twitter.com/dog_rates/status/693095443...                10   
598   https://twitter.com/dog_rates/status/683391852...                11   

      rating_denominator   name doggo floofer  pupper puppo dog_stage  
1126                  10  Ollie  None    None    None  None            
1487                  10   Milo  None    None  pupper  None    pupper  
598                   10   Jack  None    None    None  None

In [72]:
#drop other stage columns
archived_data_clean_melt.drop(['doggo', 'floofer', 'pupper', 'puppo'], axis =1, inplace=True)
archived_data_clean_melt.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name dog_stage  
0                  10  Phineas            
1                  10    Tilly

In [73]:
#view unique values in column
archived_data_clean_melt.dog_stage.unique()

array(['', 'doggo', 'puppo', 'pupper', 'floofer', 'doggopuppo',
       'doggofloofer', 'doggopupper'], dtype=object)

In [74]:
#delete rows with empty dog stage
#archived_data_clean_melt = archived_data_clean_melt[(archived_data_clean_melt.dog_stage != "")] 

In [75]:
#archived_data_clean_melt.dog_stage.unique()

In [76]:
#archived_data_clean_melt.shape

**Test**

In [77]:
archived_data_clean_melt['dog_stage'].value_counts()

                1976
pupper           245
doggo             83
puppo             29
doggopupper       12
floofer            9
doggopuppo         1
doggofloofer       1
Name: dog_stage, dtype: int64

In [78]:
archived_data_clean_melt.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name dog_stage  
0                  10  Phineas            
1                  10    Tilly

<font color='blue'> **QUALITY ISSUE - 1** </font>

**Define**

Some dog names recorded as `a` or `an` probably an error due to missign data. Rename as `None`

**Code**

In [97]:
archived_data_clean = archived_data_clean_melt

In [99]:
archived_data_clean.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name dog_stage  
0                  10  Phineas            
1                  10    Tilly

In [100]:
archived_data_clean.query('name == "a"');

In [101]:
archived_data_clean.query('name == "an"');

In [102]:
#rename rows with name as 'a' and 'an' to None
archived_data_clean.loc[archived_data_clean['name'] == "a", 'name'] = 'None'
archived_data_clean.loc[archived_data_clean['name'] == "an", 'name'] = 'None'

**Test**

In [103]:
archived_data_clean.query('name == "an"')

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, dog_stage]
Index: []

In [104]:
archived_data_clean.query('name == "a"')

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, dog_stage]
Index: []

<font color='blue'> **QUALITY ISSUE - 2** </font>

**Define**

In-consistent `rating_denominator`. Denominator cannot be greater than or less than **10**. Select rows and drop

**Code**

In [105]:
less_gret_ten = list(archived_data_clean.query('rating_denominator !=10').index)
archived_data_clean.drop(index=less_gret_ten, inplace = True)

In [106]:
archived_data_clean.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name dog_stage  
0                  10  Phineas            
1                  10    Tilly

In [107]:
archived_data_clean.rating_denominator.unique()

array([10], dtype=int64)

In [108]:
#archived_data_clean.drop(archived_data_clean.index[archived_data_clean['rating_denominator'] > 10])

**TEST**

In [109]:
archived_data_clean.rating_denominator.unique()

array([10], dtype=int64)

In [110]:
archived_data_clean.rating_denominator.value_counts()

10    2333
Name: rating_denominator, dtype: int64

<font color='blue'> **QUALITY ISSUE - 3** </font>

**DEFINE**

- Missing data for in_reply_to_status_id and in_reply_to_user_id has only 78 rows available out of the 2356. Drop these columns.

In [111]:
archived_data_clean.in_reply_to_user_id.count()

73

In [112]:
archived_data_clean.in_reply_to_user_id	.isnull().sum(), archived_data_clean.in_reply_to_status_id.isnull().sum()

(2260, 2260)

**CODE**

In [113]:
#drop columns
archived_data_clean = archived_data_clean.drop(['in_reply_to_status_id','in_reply_to_user_id'], axis=1)

**TEST**

In [114]:
archived_data_clean.head(2)

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name dog_stage  
0                  10  Phineas            
1                  10    Tilly

<font color='blue'> **QUALITY ISSUE - 4** </font>

**DEFINE**

- Not original tweet, columns retweeted_status_timestamp, retweeted status_id and user_id(drop columns). This shows rows are retweets and not original posts, retweets are not of interest in this data.

**CODE**

In [115]:
archived_data_clean.retweeted_status_id.isnull().sum(), archived_data_clean.retweeted_status_user_id.isnull().sum()

(2153, 2153)

In [116]:
#drop columns
archived_data_clean = archived_data_clean.drop(['retweeted_status_id','retweeted_status_user_id', 'retweeted_status_timestamp'], axis=1)

**TEST**

In [117]:
archived_data_clean.head(2)

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name dog_stage  
0                  10  Phineas            
1                  10    Tilly

<font color='blue'> **QUALITY ISSUE - 5** </font>

**DEFINE**
- Timestamp column is stored as object, convert to datetime

**CODE**

In [118]:
archived_data_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2333 entries, 0 to 2355
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   tweet_id            2333 non-null   int64 
 1   timestamp           2333 non-null   object
 2   source              2333 non-null   object
 3   text                2333 non-null   object
 4   expanded_urls       2278 non-null   object
 5   rating_numerator    2333 non-null   int64 
 6   rating_denominator  2333 non-null   int64 
 7   name                2333 non-null   object
 8   dog_stage           2333 non-null   object
dtypes: int64(3), object(6)
memory usage: 182.3+ KB


In [119]:
archived_data_clean.timestamp = archived_data_clean.timestamp.astype('datetime64')

**TEST**

In [272]:
archived_data_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   tweet_id            2356 non-null   int64         
 1   timestamp           2356 non-null   datetime64[ns]
 2   source              2356 non-null   object        
 3   text                2356 non-null   object        
 4   expanded_urls       2297 non-null   object        
 5   rating_numerator    2356 non-null   int64         
 6   rating_denominator  2356 non-null   int64         
 7   name                2356 non-null   object        
 8   doggo               2356 non-null   object        
 9   floofer             2356 non-null   object        
 10  pupper              2356 non-null   object        
 11  puppo               2356 non-null   object        
dtypes: datetime64[ns](1), int64(3), object(8)
memory usage: 221.0+ KB


<font color='blue'> **QUALITY ISSUE - 6** </font>

**DEFINE**
- expanded url column has multiple url's seperated by comma.
- Split expanded url column to contain only the first url

In [120]:
archived_data_clean['expanded_urls'] = archived_data_clean['expanded_urls'].str.split(',', expand=True)[0]

In [121]:
archived_data_clean.head(2)

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   
1  892177421306343426 2017-08-01 00:17:27   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name dog_stage  
0                  10  Phineas            
1                  10    Tilly

In [122]:
archived_data_clean['expanded_urls'] = archived_data_clean['expanded_urls'].str.replace(r',', '')

In [124]:
#rename to remove s from urls
archived_data_clean = archived_data_clean.rename(index=str, columns={"expanded_urls": "expanded_url"})

In [126]:
archived_data_clean.head(2)

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   
1  892177421306343426 2017-08-01 00:17:27   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   

                                        expanded_url  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name dog_stage  
0                  10  Phineas            
1                  10    Tilly

**CLEANING FOR IMAGE PREDICTION DATA**

<font color='red'> **TIDINESS ISSUE** </font>

In [133]:
image_pred_clean = image_pred
image_pred_clean.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

**DEFINE**
- Predictions, Confidence intervals and algorithm tests are stored in three columns each. Melt all three into two columns (breed and confidence)

**CODE**

In [134]:
#melt columns for the prediction number p1, p2,p3 and convert to prediction number and predicted breed
image_pred_clean = pd.melt(image_pred_clean, id_vars=['tweet_id', 'jpg_url', 'img_num', 'p1_conf', 'p1_dog', 'p2_dog', 'p3_conf', 'p3_dog'], var_name = 'pred_number',
                          value_name = 'predicted_breed')
#melt the p1_conf to p3_conf for the confidence interval of the predictions
image_pred_clean = pd.melt(image_pred_clean, id_vars=['tweet_id', 'jpg_url', 'img_num', 'p1_dog', 'p2_dog', 'p3_dog', 'pred_number', 'predicted_breed'], var_name = 'conf',
                          value_name = 'conf_interval')
#drop duplicated rows
image_pred_clean = image_pred_clean[image_pred_clean['pred_number'] == image_pred_clean['conf'].str[:2]]

#melt p1_dog to p3_dog for the accuracy of the prediction, if prediction is dog or not
image_pred_clean = pd.melt(image_pred_clean, id_vars=['tweet_id', 'jpg_url', 'img_num', 'pred_number', 'predicted_breed','conf', 'conf_interval'], var_name = 'dog_pred_num',
                          value_name = 'dog_prediction')
#remove duplicates
image_pred_clean = image_pred_clean[image_pred_clean['pred_number'] == image_pred_clean['dog_pred_num'].str[:2]]

image_pred_clean

tweet_id                                          jpg_url  \
0      666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1      666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2      666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3      666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4      666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                   ...                                              ...   
12445  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
12446  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
12447  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
12448  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
12449  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

       img_num pred_number              predicted_breed     conf  \
0            1          p1       Welsh_springer_spaniel  p1_conf   
1            1          p1                      redbone  p1_conf   
2            1          p1              German_shepherd  p1_conf   
3            1          p1          Rhodesian_ridgeback  p1_conf   
4            1          p1           miniature_pinscher  p1_conf   
...        ...         ...                          ...      ...   
12445        2          p3  German_short-haired_pointer  p3_conf   
12446        1          p3                      spatula  p3_conf   
12447        1          p3                       kelpie  p3_conf   
12448        1          p3                     papillon  p3_conf   
12449        1          p3                       banana  p3_conf   

       conf_interval dog_pred_num  dog_prediction  
0           0.465074       p1_dog            True  
1           0.506826       p1_dog            True  
2           0.596461       p1_dog            True  
3           0.408143       p1_dog            True  
4           0.560311       p1_dog            True  
...              ...          ...             ...  
12445       0.175219       p3_dog            True  
12446       0.040836       p3_dog           False  
12447       0.031379       p3_dog            True  
12448       0.068957       p3_dog            True  
12449       0.076110       p3_dog           False  

[4150 rows x 9 columns]

In [135]:
image_pred.shape, image_pred_clean.shape

((2075, 12), (4150, 9))

In [136]:
#drop unnecessary columns
image_pred_clean = image_pred_clean.drop(['conf','dog_pred_num'], axis=1)

In [131]:
image_pred_clean

tweet_id                                          jpg_url  \
0      666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1      666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2      666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3      666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4      666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                   ...                                              ...   
12445  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
12446  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
12447  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
12448  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
12449  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

       img_num pred_number              predicted_breed  conf_interval  \
0            1          p1       Welsh_springer_spaniel       0.465074   
1            1          p1                      redbone       0.506826   
2            1          p1              German_shepherd       0.596461   
3            1          p1          Rhodesian_ridgeback       0.408143   
4            1          p1           miniature_pinscher       0.560311   
...        ...         ...                          ...            ...   
12445        2          p3  German_short-haired_pointer       0.175219   
12446        1          p3                      spatula       0.040836   
12447        1          p3                       kelpie       0.031379   
12448        1          p3                     papillon       0.068957   
12449        1          p3                       banana       0.076110   

       dog_prediction  
0                True  
1                True  
2                True  
3                True  
4                True  
...               ...  
12445            True  
12446           False  
12447            True  
12448            True  
12449           False  

[4150 rows x 7 columns]

In [137]:
image_pred_clean.duplicated().sum().any()

False

In [138]:
image_pred_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4150 entries, 0 to 12449
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   tweet_id         4150 non-null   int64  
 1   jpg_url          4150 non-null   object 
 2   img_num          4150 non-null   int64  
 3   pred_number      4150 non-null   object 
 4   predicted_breed  4150 non-null   object 
 5   conf_interval    4150 non-null   float64
 6   dog_prediction   4150 non-null   bool   
dtypes: bool(1), float64(1), int64(2), object(3)
memory usage: 231.0+ KB


<font color='blue'> **QUALITY ISSUE - 1** </font>

**DEFINE**

Prediction number should be an integer. remove p from pred_number column and change to int64

**CODE**

In [139]:
image_pred_clean['pred_number'] = image_pred_clean['pred_number'].str.replace(r'\D', '').astype(int)

In [535]:
image_pred_clean

tweet_id                                          jpg_url  \
0      666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1      666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2      666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3      666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4      666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                   ...                                              ...   
12445  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
12446  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
12447  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
12448  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
12449  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

       img_num  pred_number              predicted_breed  conf_interval  \
0            1            1       Welsh_springer_spaniel       0.465074   
1            1            1                      redbone       0.506826   
2            1            1              German_shepherd       0.596461   
3            1            1          Rhodesian_ridgeback       0.408143   
4            1            1           miniature_pinscher       0.560311   
...        ...          ...                          ...            ...   
12445        2            3  German_short-haired_pointer       0.175219   
12446        1            3                      spatula       0.040836   
12447        1            3                       kelpie       0.031379   
12448        1            3                     papillon       0.068957   
12449        1            3                       banana       0.076110   

       dog_prediction  
0                True  
1                True  
2                True  
3                True  
4                True  
...               ...  
12445            True  
12446           False  
12447            True  
12448            True  
12449           False  

[4150 rows x 7 columns]

In [140]:
image_pred_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4150 entries, 0 to 12449
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   tweet_id         4150 non-null   int64  
 1   jpg_url          4150 non-null   object 
 2   img_num          4150 non-null   int64  
 3   pred_number      4150 non-null   int32  
 4   predicted_breed  4150 non-null   object 
 5   conf_interval    4150 non-null   float64
 6   dog_prediction   4150 non-null   bool   
dtypes: bool(1), float64(1), int32(1), int64(2), object(2)
memory usage: 214.8+ KB


<font color='blue'> **QUALITY ISSUE - 2** </font>

**DEFINE**
- inconsistent naming for dog breeds p1. Remove `_` from breed name

**CODE**

In [141]:
#remove **_** from dog breed prediction
image_pred_clean['predicted_breed'] = image_pred_clean['predicted_breed'].str.replace(r'_', ' ')

**TEST**

In [143]:
image_pred_clean.head(2)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   

   img_num  pred_number         predicted_breed  conf_interval  dog_prediction  
0        1            1  Welsh springer spaniel       0.465074            True  
1        1            1                 redbone       0.506826            True

<font color='blue'> **QUALITY ISSUE - 3** </font>

**DEFINE**
- inconsistent labelling for dog breeds, convert all breed name to lower case

**CODE**

In [144]:
#convert breeds to lower case
image_pred_clean['predicted_breed'] = image_pred_clean['predicted_breed'].str.lower()

In [146]:
image_pred_clean.head(2)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   

   img_num  pred_number         predicted_breed  conf_interval  dog_prediction  
0        1            1  welsh springer spaniel       0.465074            True  
1        1            1                 redbone       0.506826            True

<font color='blue'> **QUALITY ISSUE - 4** </font>

**DEFINE**
- Some dog predictins are false, filter out rows

In [147]:
image_pred_clean.dog_prediction.value_counts()

True     3031
False    1119
Name: dog_prediction, dtype: int64

**CODE**

In [149]:
image_pred_clean = image_pred_clean[(image_pred_clean[['dog_prediction']] != False).all(axis=1)]

**TEST**

In [151]:
image_pred_clean.query('dog_prediction == False')

Empty DataFrame
Columns: [tweet_id, jpg_url, img_num, pred_number, predicted_breed, conf_interval, dog_prediction]
Index: []

**CLEANING RETWEET AND FAVORITE COUNT DATA**

**DEFINE**

- DELETE first column(unamed 0) not needed.

**CODE**

In [152]:
rtwt_fav_count.head(2)

Unnamed: 0            tweet_id  retweet_count  favourite_count
0           0  892420643555336193           7340            34978
1           1  892177421306343426           5475            30279

In [153]:
#drop unnamed column
rtwt_fav_count.drop('Unnamed: 0', axis=1, inplace=True)

**TEST**

In [155]:
rtwt_fav_count.head(2)

tweet_id  retweet_count  favourite_count
0  892420643555336193           7340            34978
1  892177421306343426           5475            30279

#### CREATE MASTER `.csv` FOR THE THREE DATAFRAMES

In [157]:
image_pred_clean.head()
rtwt_fav_count.head()
archived_data_clean.head();

In [ ]:
image_pred_clean.to_csv('image_predictions.csv', sep=',', index=False)
rtwt_fav_count.to_csv('retweets_counts.csv', sep=',', index=False)
archived_data_clean.to_csv('archived_data.csv', sep=',', index=False)

In [158]:
#merge the three datasets
twitter_archive_master = pd.merge(archived_data_clean, rtwt_fav_count, how='left', on=['tweet_id'])
twitter_archive_master = pd.merge(twitter_archive_master, image_pred_clean, how='inner', on=['tweet_id'])

Due to the difference in the total number of each dataset, there was a few gaps and `NaN` ocured in few of the rows.

In [159]:
twitter_archive_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3001 entries, 0 to 3000
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   tweet_id            3001 non-null   int64         
 1   timestamp           3001 non-null   datetime64[ns]
 2   source              3001 non-null   object        
 3   text                3001 non-null   object        
 4   expanded_url        3001 non-null   object        
 5   rating_numerator    3001 non-null   int64         
 6   rating_denominator  3001 non-null   int64         
 7   name                3001 non-null   object        
 8   dog_stage           3001 non-null   object        
 9   retweet_count       2977 non-null   float64       
 10  favourite_count     2977 non-null   float64       
 11  jpg_url             3001 non-null   object        
 12  img_num             3001 non-null   int64         
 13  pred_number         3001 non-null   int32       

After merging the datasets, probably due to the difference in the number of rows the `retweet_count and favorite_counts` has some `NaN values.` But according to this response in [Stackoverflow](https://stackoverflow.com/questions/21287624/convert-pandas-column-containing-nans-to-dtype-int) the best option is to work with `floats` hence I didnt convert them to int or remove the rows with missing values.

In [160]:
twitter_archive_master.shape, twitter_archive_master.isnull().sum()

((3001, 17),
 tweet_id               0
 timestamp              0
 source                 0
 text                   0
 expanded_url           0
 rating_numerator       0
 rating_denominator     0
 name                   0
 dog_stage              0
 retweet_count         24
 favourite_count       24
 jpg_url                0
 img_num                0
 pred_number            0
 predicted_breed        0
 conf_interval          0
 dog_prediction         0
 dtype: int64)

#### STORE DATA

The merged data was stored as a `.csv` in my local storage. 

In [ ]:
twitter_archive_master.to_csv('twitter_archive_master.csv', sep=',', index=False)

### VISUALISATION AND ANALYSIS

The wrangling part has been concluded. the next step is that the wrangled data will be analysed and visualised here. This should include at least three (3) insights and one (1) visualisation.

**Possible questions I attempted to answer in this section:**

- The most common dog breed
- Which dog is the most retweeted and favorite dog
- What is the most common or popular dog stage
- Which dog stage has the best ratings

### RESOURCES

* https://www.earthdatascience.org/courses/use-data-open-source-python/intro-to-apis/twitter-data-in-python/
* https://towardsdatascience.com/data-analysis-project-weratedogs-16aad3f6c8c2
